### PR TITLE
Replace floating point from_chars with newer fast_float

### LINKED ISSUE #55: result's ptr is not correct for some numbers
This fails:
```c++
string_view s = "0e0";
auto err = detail::charconv::from_chars( s.begin(), s.end(), d );
BOOST_ASSERT( err.ptr == s.end() );
```

Same for e.g. `0e00000000000` and `0e1`.

### PR DESCRIPTION
**Pull request by mborland**
See: https://github.com/fastfloat/fast_float
As of version 5 it is licensed under BSL.

### CODE REVIEW

**Comment by mborland** [comment]
Here are the benchmarks from `benchmarks/from_chars_floating.cpp`:

On current develop:
```
GNU C++ version 13.1.0
GNU libstdc++ version 20230426

                std::strtox<float>, scientific:   755 ms (s=-2.83944e+20)
            std::from_chars<float>, scientific:   731 ms (s=-2.83944e+20)
boost::charconv::from_chars<float>, scientific:  2696 ms (s=-2.83944e+20)

                std::strtox<double>, scientific:   919 ms (s=1.47618e+299)
            std::from_chars<double>, scientific:   715 ms (s=1.47618e+299)
boost::charconv::from_chars<double>, scientific:  3724 ms (s=1.47618e+299)

                std::strtox<float>, general:   746 ms (s=-2.83944e+20)
            std::from_chars<float>, general:   704 ms (s=-2.83944e+20)
boost::charconv::from_chars<float>, general:  2620 ms (s=-2.83944e+20)

                std::strtox<double>, general:   912 ms (s=1.47618e+299)
            std::from_chars<double>, general:   713 ms (s=1.47618e+299)
boost::charconv::from_chars<double>, general:  3898 ms (s=1.47618e+299)

                std::strtox<float>, uint64:   788 ms (s=1.01167e+19)
            std::from_chars<float>, uint64:   916 ms (s=1.01167e+19)
boost::charconv::from_chars<float>, uint64:  3481 ms (s=1.01167e+19)

                std::strtox<double>, uint64:   767 ms (s=1.01167e+19)
            std::from_chars<double>, uint64:   912 ms (s=1.01167e+19)
boost::charconv::from_chars<double>, uint64:  3294 ms (s=1.01167e+19)
```

With this PR:

```
GNU C++ version 13.1.0
GNU libstdc++ version 20230426

                std::strtox<float>, scientific:   758 ms (s=-2.83944e+20)
            std::from_chars<float>, scientific:   730 ms (s=-2.83944e+20)
boost::charconv::from_chars<float>, scientific:  2306 ms (s=-2.83944e+20)

                std::strtox<double>, scientific:   917 ms (s=1.47618e+299)
            std::from_chars<double>, scientific:   714 ms (s=1.47618e+299)
boost::charconv::from_chars<double>, scientific:  2441 ms (s=1.47618e+299)

                std::strtox<float>, general:   747 ms (s=-2.83944e+20)
            std::from_chars<float>, general:   705 ms (s=-2.83944e+20)
boost::charconv::from_chars<float>, general:  2208 ms (s=-2.83944e+20)

                std::strtox<double>, general:   909 ms (s=1.47618e+299)
            std::from_chars<double>, general:   710 ms (s=1.47618e+299)
boost::charconv::from_chars<double>, general:  2448 ms (s=1.47618e+299)

                std::strtox<float>, uint64:   749 ms (s=1.01167e+19)
            std::from_chars<float>, uint64:   869 ms (s=1.01167e+19)
boost::charconv::from_chars<float>, uint64:  3149 ms (s=1.01167e+19)

                std::strtox<double>, uint64:   734 ms (s=1.01167e+19)
            std::from_chars<double>, uint64:   871 ms (s=1.01167e+19)
boost::charconv::from_chars<double>, uint64:  3150 ms (s=1.01167e+19)
```

**Comment by mborland** [comment]
Benchmarks from `benchmarks/from_chars_floating.cpp`. Run on an M1 Mac with macOS 13.4 and GCC 13:
```
GNU C++ version 13.1.0
GNU libstdc++ version 20230426

                std::strtox<float>, scientific:   607 ms (s=-2.83944e+20)
            std::from_chars<float>, scientific:   530 ms (s=-2.83944e+20)
boost::charconv::from_chars<float>, scientific:   535 ms (s=-2.83944e+20)

                std::strtox<double>, scientific:   778 ms (s=1.47618e+299)
            std::from_chars<double>, scientific:   531 ms (s=1.47618e+299)
boost::charconv::from_chars<double>, scientific:   529 ms (s=1.47618e+299)

                std::strtox<float>, general:   597 ms (s=-2.83944e+20)
            std::from_chars<float>, general:   513 ms (s=-2.83944e+20)
boost::charconv::from_chars<float>, general:   519 ms (s=-2.83944e+20)

                std::strtox<double>, general:   770 ms (s=1.47618e+299)
            std::from_chars<double>, general:   528 ms (s=1.47618e+299)
boost::charconv::from_chars<double>, general:   526 ms (s=1.47618e+299)

                std::strtox<float>, uint64:   571 ms (s=1.01167e+19)
            std::from_chars<float>, uint64:   703 ms (s=1.01167e+19)
boost::charconv::from_chars<float>, uint64:   713 ms (s=1.01167e+19)

                std::strtox<double>, uint64:   564 ms (s=1.01167e+19)
            std::from_chars<double>, uint64:   702 ms (s=1.01167e+19)
boost::charconv::from_chars<double>, uint64:   708 ms (s=1.01167e+19)
```

On develop:

```
GNU C++ version 13.1.0
GNU libstdc++ version 20230426

                std::strtox<float>, scientific:   607 ms (s=-2.83944e+20)
            std::from_chars<float>, scientific:   528 ms (s=-2.83944e+20)
boost::charconv::from_chars<float>, scientific:   763 ms (s=-2.83944e+20)

                std::strtox<double>, scientific:   780 ms (s=1.47618e+299)
            std::from_chars<double>, scientific:   527 ms (s=1.47618e+299)
boost::charconv::from_chars<double>, scientific:  1024 ms (s=1.47618e+299)

                std::strtox<float>, general:   594 ms (s=-2.83944e+20)
            std::from_chars<float>, general:   512 ms (s=-2.83944e+20)
boost::charconv::from_chars<float>, general:   745 ms (s=-2.83944e+20)

                std::strtox<double>, general:   770 ms (s=1.47618e+299)
            std::from_chars<double>, general:   529 ms (s=1.47618e+299)
boost::charconv::from_chars<double>, general:  1024 ms (s=1.47618e+299)

                std::strtox<float>, uint64:   585 ms (s=1.01167e+19)
            std::from_chars<float>, uint64:   705 ms (s=1.01167e+19)
boost::charconv::from_chars<float>, uint64:   793 ms (s=1.01167e+19)

                std::strtox<double>, uint64:   566 ms (s=1.01167e+19)
            std::from_chars<double>, uint64:   706 ms (s=1.01167e+19)
boost::charconv::from_chars<double>, uint64:   782 ms (s=1.01167e+19)
```

**Comment by mborland** [comment]
@pdimov I think this is good for review. The fast_float implementation is generally untouched outside of changing some macros.

File: include/boost/charconv/chars_format.hpp [hunk 1]
```diff
@@ -5,11 +5,13 @@
...
 #define BOOST_CHARCONV_CHARS_FORMAT_HPP
 
+#include <boost/charconv/detail/config.hpp>
+
 namespace boost { namespace charconv {
 
 // Floating-point format for primitive numerical conversion
 // chars_format is a bitmask type (16.3.3.3.3)
-enum class chars_format : unsigned
+enum chars_format : unsigned
```
Comment (pdimov) [inline include/boost/charconv/chars_format.hpp]: This change doesn't seem correct to me. `chars_format` is part of the interface and is specified to be `enum class`. Why is this needed?

File: include/boost/charconv/chars_format.hpp [hunk 2]
```diff
@@ -5,11 +5,13 @@
 #ifndef BOOST_CHARCONV_CHARS_FORMAT_HPP
 #define BOOST_CHARCONV_CHARS_FORMAT_HPP
 
+#include <boost/charconv/detail/config.hpp>
```
Comment (pdimov) [inline include/boost/charconv/chars_format.hpp]: Why is this include needed?

File: include/boost/charconv/chars_format.hpp (hunk 1 above)
Comment (mborland) [reply include/boost/charconv/chars_format.hpp]: He uses bitwise operators in the control flow (e.g. https://github.com/cppalliance/charconv/pull/54/files#diff-257fe44b316853ce9ec340001186ef24affd04a1406ece4f47b98a0bdb77b95dR200). I tried implementing all of the bitwise operators and then casting to bool, but it was ugly and compilers complained about forcing to bool. The unused header above is cruft from that attempt.

Comment (pdimov) [reply include/boost/charconv/chars_format.hpp]: We can't just change the interface to whatever is convenient to the current implementation. `fmt & chars_format::scientific` needs to be changed to `(unsigned)fmt & (unsigned)chars_format::scientific`.

**Pull request merged**